### ISSUE TITLE
`log_model`

### ISSUE THREAD

**Issue by dberenbaum** [iterative/dvclive#472]
Depends on #378.

Replaces #305.

#378 adds support for tracking any file or directory artifact with DVC. In #305, there was discussion of saving model-specific artifacts using MLEM. Instead of using MLEM, this issue focuses on how to track an arbitrary model file like `log_artifact`, except that it should be tagged as a model so it is automatically shown in the Studio model registry and can be viewed, promoted, consumed, etc.

cc @aguschin Let's think about this as the starting point for the basic model registry workflow.

**Comment by dberenbaum** [comment]
DVC already has some built-in support that could be useful: https://github.com/iterative/dvc/issues/8214. `log_model` could add to DVC with `type: model` for now, and Studio/GTO could consume that to show in the model registry. Could that be enough to support a simple end-to-end model workflow?

This could probably depend on the path of the model rather than a specified model name (which is what GTO uses now in `artifacts.yaml`). If that becomes too limited, we could add top-level models section that looks like GTO's `artifacts.yaml` as a follow-up.

**Comment by aguschin** [comment]
> Could that be enough to support a simple end-to-end model workflow

In general, yes. Although if we don't merge `artifacts.yaml` into DVC at that moment, we're getting two sources of info for artifacts. Now users may wonder why `gto show` doesn't show models logged with `log_model`.

We can go ahead and merge `artifacts.yaml` (which, IIUC, means taking https://github.com/iterative/dvc/issues/8214, adding back `dvc models list` implemented there already, taking out `artifacts.yaml` from GTO, updating docs and communicating that to customers and OSS users).

What's unclear to me, is removing `name` concept. In this case, it can break users workflow (now they need to keep somewhere the connection between `path` and `name` themselves, otherwise GTO and DVC won't work together in [scenarios we described](https://www.notion.so/iterative/GTO-DVC-integration-fee2d04b69c24de6a04dbac48fd99847), including "in CI, download a model called `mymodel` promoted to `prod`" scenario.

**Comment by dberenbaum** [comment]
> Now users may wonder why `gto show` doesn't show models logged with `log_model`.

Sorry, my assumption here was that `gto show` would show the models logged with `log_model`.

>     * taking [Meta information for data dvc#8214](https://github.com/iterative/dvc/issues/8214),
> 
>     * adding back `dvc models list` implemented there already,

Do you mean `dvc data ls --type models` instead of `dvc models list`?

>     * implementing metadata in pipelines,

I thought it was implemented in https://github.com/iterative/dvc/pull/8251?

> What's unclear to me, is not implementing `name` concept in DVC.

Sorry, I meant that if we default to using path when there is no `name`, we can simplify the basic scenario.

> I think it's a valid solution, but we as well could just add `name: something` field (just like `type: model`). It may work out fine even without `models: ...` section. But this may become clearer when we start implementing this for pipelines

Sounds good. Agreed that it's not clear whether we need a `models:` section.

**Comment by aguschin** [comment]
> Do you mean `dvc data ls --type models` instead of `dvc models list`?

Yes, I meant that one, thanks. But I also would consider adding simpler commands `dvc models ls` and `dvc data ls` and `dvc artifacts ls` to highlight that DVC has a specific notion of "model" and "data" now. Not sure if that's possible considering existing `dvc data status` command though.

Right, just checked - it was merged indeed. I'll update [this Notion page](https://www.notion.so/iterative/GTO-DVC-integration-fee2d04b69c24de6a04dbac48fd99847) with regard to that.

Also, side-thought: as an alternative, we could make GTO integrated with DVC for read operations. So instead of
```sh
$ REV=$(...some gto command...)
$ dvc get $REPO $MODEL_PATH --rev $REV -o $MODEL_PATH
```
one would be able to also run
```sh
$ gto get $REPO $MODEL_NAME --version v1.0.0
# or 
$ gto get $REPO $MODEL_NAME --stage prod
```
Yet not sure how good is this idea from other angles, but having one command to download models, and do that without going back-and-forth with model names, paths and revisions sounds great to me.

**Comment by dberenbaum** [comment]
@aguschin suggested an alternative to `log_model` using an argument to `live.log_artifact`:

```
live.log_artifact(artifact, "path", type="model")
```

Makes sense to me! No need for an entirely separate method here.

**Comment by dberenbaum** [comment]
@aguschin @daavoo Related to the discussions in https://github.com/iterative/mlem.ai/pull/323: should `live.log_artifact("path/to/artifact")` always add something like this to `dvclive/dvc.yaml` even if there is no type specified?

```yaml
artifacts:
  - path/to/artifact
```

Also, do we plan to add other optional fields to `log_artifact` like `name/alias`, `description`, etc.?

None of this is urgent IMO but all would be nice to have.

**Comment by aguschin** [comment]
@dberenbaum, it happened that we're discussing the same thing here https://github.com/iterative/dvc/pull/9220#discussion_r1147531254, but let me answer with my current understanding:

1. We want to write `dvc.yaml` at `live.log_artifact("path/to/artifact")`, but there are some corner cases - asking about them in the PR I linked.
2. We want to allow logging `name`, `description`, `type`, etc.

We can continue the discussion here or there, whichever works best.

**Comment by daavoo** [comment]
Moving the discussion to here.

> At this step, if myartifactname exists in any other dvc.yaml, DVC will throw an error since no names duplication is allowed. How we should handle this? Is it ok to break runtime because of this (especially if user specified type, name, etc)? Or maybe better to issue a warning and continue code execution?

In general, we don't want DVCLive to fail but just warn.
In this specific case, I am not sure if we should check for duplicates at all. IMO, checking for duplicates should happen at a different level not at writing time.

> if DVCLive wasn't set up to write to dvc.yaml the way you described, but user calls log_artifact(path, name="myname") or pass any other args (like labels, type or description), what should DVCLive do? 

I think we want to keep it simple / consistent with current behavior:

- `log_artifact(path, **kwargs)` will write to `path.dvc`, including the kwargs (`name`, `labels`, `type`, etc.).
- If `dvcyaml=True`, we will add all that info to the `artifacts` section.
- `log_model` would be just an alias for `log_artifact(path, type="model", **kwargs)`

> I guess the answer should be similar to the case when one calls log_metric, but forgets to call make_dvcyaml at the end. What happens then, could you please explain?

https://dvc.org/doc/dvclive/how-it-works

`log_{X}` calls always write to some output file.
`make_{X}` calls are called on `end` and/or `next_step` depending on arguments passed to `Live`.

The user would not really forget to call `make_dvcyaml` but rather would be called depending on the value the user has passed to `Live(dvcyaml=True/False)`.

**Comment by aguschin** [comment]
> checking for duplicates should happen at a different level, not at writing time

Then after running `log_artifact` and committing changes, user may end up with a broken DVC repo (two artifacts with the same `name`, DVC don't know which one to take into account). Doesn't sound right to me... Is there some way to prevent this?

> log_artifact(path, **kwargs) will write to path.dvc, including the kwargs (name, labels, type, etc.).

Sounds logical from dvclive/user perspective. Then as sources of `artifacts` we're having not only `artifacts:` section,  also `.dvc` files.

Question then (again): are we showing in Studio artifacts from both places?

**Comment by dberenbaum** [comment]
> Then after running `log_artifact` and committing changes, user may end up with a broken DVC repo (two artifacts with the same `name`, DVC don't know which one to take into account). Doesn't sound right to me... Is there some way to prevent this?

We also encourage manually editing `dvc.yaml`, so there's no way to prevent this. Having mechanisms to write to `dvc.yaml` and validate beforehand are nice to have but I don't think they are a priority. Let's please keep it simple to start, because I worry this will delay this PR which otherwise seems close to being ready to merge.

> Question then (again): are we showing in Studio artifacts from both places?

My take from https://github.com/iterative/mlem.ai/pull/323 was that the `artifacts:` section was the sole source of truth for GTO and the Studio model registry. We may want to drop support for `desc`, `type`, `labels` from `.dvc` files, although I don't think it's an immediate priority.

**Comment by shcheklein** [comment]
Folks, is this one done?
